pkg/cloudprovider/servicecontroller/servicecontroller.go

```diff
@@ -0,0 +1,468 @@
+		cloud:       cloud,
+		kubeClient:  kubeClient,
+		clusterName: clusterName,
+		cache:       &serviceCache{serviceMap: make(map[string]*cachedService)},
+	}
+}
+
+// Run starts a background goroutine that watches for changes to services that
+// have (or had) externalLoadBalancers=true and ensures that they have external
+// load balancers created and deleted appropriately.
+func (s *ServiceController) Run() error {
+	if err := s.init(); err != nil {
+		return err
+	}
+
+	// We have to make this check beecause the ListWatch that we use in
+	// WatchServices requires Client functions that aren't in the interface
+	// for some reason.
+	if _, ok := s.kubeClient.(*client.Client); !ok {
+		return fmt.Errorf("ServiceController only works with real Client objects, but was passed something else satisfying the client Interface.")
+	}
+
+	// Get the currently existing set of services and then all future creates
+	// and updates of services.
+	// No delta compressor is needed for the DeltaFIFO queue because we only ever
+	// care about the most recent state.
+	serviceQueue := cache.NewDeltaFIFO(cache.MetaNamespaceKeyFunc, nil, s.cache)
+	lw := cache.NewListWatchFromClient(s.kubeClient.(*client.Client), "services", api.NamespaceAll, fields.Everything())
+	cache.NewReflector(lw, &api.Service{}, serviceQueue, 0).Run()
+	for i := 0; i < workerGoroutines; i++ {
+		go s.watchServices(serviceQueue)
+	}
+	return nil
+}
+
+func (s *ServiceController) init() error {
+	if s.cloud == nil {
+		return fmt.Errorf("ServiceController should not be run without a cloudprovider.")
+	}
+
+	balancer, ok := s.cloud.TCPLoadBalancer()
+	if !ok {
+		return fmt.Errorf("the cloud provider does not support external TCP load balancers.")
+	}
+	s.balancer = balancer
+
+	zones, ok := s.cloud.Zones()
+	if !ok {
+		return fmt.Errorf("the cloud provider does not support zone enumeration, which is required for creating external load balancers.")
+	}
+	zone, err := zones.GetZone()
+	if err != nil {
+		return fmt.Errorf("failed to get zone from cloud provider, will not be able to create external load balancers: %v", err)
+	}
+	s.zone = zone
+	return nil
+}
+
+// Loop infinitely, processing all service updates provided by the queue.
+func (s *ServiceController) watchServices(serviceQueue *cache.DeltaFIFO) {
+	for {
+		newItem := serviceQueue.Pop()
+		deltas, ok := newItem.(cache.Deltas)
+		if !ok {
+			glog.Errorf("Received object from service watcher that wasn't Deltas: %+v", newItem)
+		}
+		delta := deltas.Newest()
+		if delta == nil {
+			glog.Errorf("Received nil delta from watcher queue.")
+			continue
+		}
+		err, shouldRetry := s.processDelta(delta)
+		if shouldRetry {
+			// Add the failed service back to the queue so we'll retry it.
+			glog.Errorf("Failed to process service delta. Retrying: %v", err)
+			time.Sleep(5 * time.Second)
+			serviceQueue.AddIfNotPresent(deltas)
+		} else if err != nil {
+			glog.Errorf("Failed to process service delta. Not retrying: %v", err)
```

Comment:
You're still just silently logging errors here. Short of digging through logs and noticing the error, the user won't know that an error has occurred, or that you've given up trying to build a working service for her. I think that you need to have the service go into some sort of error state in this case.